src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/PrototypeParser.cs

```diff
@@ -0,0 +1,586 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO.Pipelines;
+using System.Numerics;
+using System.Runtime.CompilerServices;
+using System.Text.Utf8;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.Http;
+using Microsoft.AspNetCore.Server.Kestrel.Internal.Infrastructure;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Internal.Http
+{
+    public class PrototypeParser : IHttpParser
+    {
+        public PrototypeParser(IKestrelTrace log)
+        {
+            Log = log;
+        }
+
+        private IKestrelTrace Log { get; }
+
+        // byte types don't have a data type annotation so we pre-cast them; to avoid in-place casts
+        private const byte ByteCR = (byte)'\r';
+        private const byte ByteLF = (byte)'\n';
+        private const byte ByteColon = (byte)':';
+        private const byte ByteSpace = (byte)' ';
+        private const byte ByteTab = (byte)'\t';
+        private const byte ByteQuestionMark = (byte)'?';
+        private const byte BytePercentage = (byte)'%';
+
+        public unsafe bool ParseRequestLine<T>(T handler, ReadableBuffer buffer, out ReadCursor consumed, out ReadCursor examined) where T : IHttpRequestLineHandler
+        {
+            consumed = buffer.Start;
+            examined = buffer.End;
+
+            var start = buffer.Start;
+            if (ReadCursorOperations.Seek(start, buffer.End, out var end, ByteLF) == -1)
+            {
+                return false;
+            }
+
+            // Move 1 byte past the \n
+            end = buffer.Move(end, 1);
+            var startLineBuffer = buffer.Slice(start, end);
+
+            Span<byte> span;
+            if (startLineBuffer.IsSingleSpan)
+            {
+                // No copies, directly use the one and only span
+                span = startLineBuffer.First.Span;
+            }
+            else
+            {
+                // We're not a single span here but we can use pooled arrays to avoid allocations in the rare case
+                span = new Span<byte>(new byte[startLineBuffer.Length]);
+                startLineBuffer.CopyTo(span);
+            }
+
+            var pathStart = -1;
+            var queryStart = -1;
+            var queryEnd = -1;
+            var pathEnd = -1;
+            var versionStart = -1;
+
+            HttpVersion httpVersion = HttpVersion.Unknown;
+            HttpMethod method;
+            Span<byte> customMethod;
+            int i = 0;
+            var length = span.Length;
+            var done = false;
+
+            fixed (byte* data = &span.DangerousGetPinnableReference())
+            {
+                switch (StartLineState.KnownMethod)
+                {
+                    case StartLineState.KnownMethod:
+                        if (span.GetKnownMethod(out method, out var methodLength))
```

Comment:
Why is this inside switch/case? The beginning of the request line has to be the method/verb and then it will never be method/verb, and so the case will execute once at the beginning and never again. It means it can be moved out of the switch block.